File: BinarySearch/minEatingSpeed.js
const minEatingSpeed = (piles, h) => {
  const check = k => {
    let totalHours = 0;
    for (const bananas of piles) {
      totalHours += Math.ceil(bananas / k);
      if (totalHours > h) {
        return false;
      }
    }
    return true;
  }

  let left = 1
  let right = Math.max(...piles);
  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (check(mid)) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;
}

// Time complexity: O(n log m) where n is the number of piles and m is the
// maximum pile size.
// Space complexity: O(1)

File: BinarySearch/MinEatingSpeed.java
class MinEatingSpeed {
  public int minEatingSpeed(int[] piles, int h) {
    int left = 1;
    int right = 0;
    for (int pile : piles)
      right = Math.max(right, pile);
    while (left <= right) {
      int mid = left + (right - left) / 2;
      if (check(mid, piles, h)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }
    return left;
  }

  private boolean check(double eatingSpeed, int[] piles, int h) {
    long totalHoursToEat = 0;
    for (int pile : piles) {
      totalHoursToEat += (pile + eatingSpeed - 1) / eatingSpeed;
      if (totalHoursToEat > h) {
        return false;
      }
    }

    return totalHoursToEat <= h;
  }
}

// Time complexity: O(n log m) where n is the number of piles and m is the
// maximum pile size.
// Space complexity: O(1)